### PR TITLE
Add connect/disconnect on Tie Lines and HVDC Lines

### PR DESCRIPTION
**Pull request by rolnico**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**
No

**What kind of change does this PR introduce?**
Bug fix + new Feature

**What is the current behavior?**
If you use for example a ConnectableConnection on a non-existing connectable or on a TieLine (which is not a Connectable), the application of the NetworkModification will crash due to a null pointer exception. This is due to the fact that only Connectables are managed since Tie Lines and HVDC Lines do not have connect/disconnect methods.

**What is the new behavior (if this is a feature change)?**
Tie Lines and HVDC Lines now have connect and disconnect methods implemented.
Network modification for connection and disconnection now also work with Tie Lines and HVDC Lines.

**Does this PR introduce a breaking change or deprecate an API?**
- [X] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [X] The *Breaking Change* or *Deprecated* label has been added
- [X] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

If you had your own implementation for HVDC Lines, you now need to implement the following methods:
```java
    boolean connectConverterStations();
    boolean connectConverterStations(Predicate<Switch> isTypeSwitchToOperate);
    boolean connectConverterStations(Predicate<Switch> isTypeSwitchToOperate, TwoSides side);
    boolean disconnectConverterStations();
    boolean disconnectConverterStations(Predicate<Switch> isSwitchOpenable);
    boolean disconnectConverterStations(Predicate<Switch> isSwitchOpenable, TwoSides side);
```

If you had your own implementation for Tie Lines, you now need to implement the following methods:
```java
    boolean connectDanglingLines();
    boolean connectDanglingLines(Predicate<Switch> isTypeSwitchToOperate);
    boolean connectDanglingLines(Predicate<Switch> isTypeSwitchToOperate, TwoSides side);
    boolean disconnectDanglingLines();
    boolean disconnectDanglingLines(Predicate<Switch> isSwitchOpenable);
    boolean disconnectDanglingLines(Predicate<Switch> isSwitchOpenable, TwoSides side);
```

### CODE REVIEW

**Comment by rolnico** [comment]
This is another possible solution that would close https://github.com/powsybl/powsybl-core/pull/3069

**Comment by rolnico** [comment]
Should the Tie Line and HVDC Line become Connectables ?

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/HvdcLine.java [hunk 1]
```diff
@@ -204,6 +207,20 @@ default HvdcConverterStation<?> getConverterStation(TwoSides side) {
...
+
+    boolean connect(Predicate<Switch> isTypeSwitchToOperate, ThreeSides side);
+
+    boolean disconnect();
+
+    boolean disconnect(Predicate<Switch> isSwitchOpenable);
+
+    boolean disconnect(Predicate<Switch> isSwitchOpenable, ThreeSides side);
+
+    List<Terminal> getTerminals(ThreeSides side);
```
Comment (flo-dup) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/HvdcLine.java:222]: We don't need to expose `getTerminals` so you could remove it here

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/HvdcLine.java [hunk 2]
```diff
@@ -204,6 +207,20 @@ default HvdcConverterStation<?> getConverterStation(TwoSides side) {
      */
     void remove();
 
+    boolean connect();
+
+    boolean connect(Predicate<Switch> isTypeSwitchToOperate);
+
+    boolean connect(Predicate<Switch> isTypeSwitchToOperate, ThreeSides side);
```
Comment (flo-dup) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/HvdcLine.java:214]: `TwoSides` would make more sense here and in disconnect

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TieLine.java
Comment (flo-dup) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TieLine.java:1]: Same comments for TieLine

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/HvdcLineImpl.java [hunk 3]
```diff
@@ -231,6 +241,145 @@ public void remove() {
         network.getListeners().notifyAfterRemoval(id);
     }
 
+    @Override
+    public boolean connect() {
+        return connect(SwitchPredicates.IS_NONFICTIONAL_BREAKER);
```
Comment (flo-dup) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/HvdcLineImpl.java:246]: didn't realize, but isn't it fictitious rather than fictional?

**Review by flo-dup — CHANGES_REQUESTED** [review]
Could you add a utility method to avoid the code duplication?

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/HvdcLineImpl.java (hunk 3 above)
Comment (rolnico) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/HvdcLineImpl.java]: both can be used, IMO.
If we want to change it, we should do it in another PR

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractConnectable.java [hunk 4]
```diff
@@ -206,59 +205,13 @@ public boolean connect(Predicate<Switch> isTypeSwitchToOperate) {
...
-
-        // Connect all bus-breaker terminals
-        for (TerminalExt terminal : getTerminals(side)) {
-            if (!terminal.isConnected()
-                && terminal.getVoltageLevel().getTopologyKind() == BUS_BREAKER) {
-                // At this point, isNowConnected should always stay true but let's be careful
-                isNowConnected = isNowConnected && terminal.connect(isTypeSwitchToOperate);
-            }
-        }
+        List<Terminal> terminalsToConnect = new ArrayList<>(getTerminals(side));
```
Comment (flo-dup) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractConnectable.java:208]: I don't think you need to copy the list, you only iterate over it in the connect. You did it for the case of the connectable being removed while connecting it?

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractConnectable.java [hunk 5]
```diff
@@ -206,59 +205,13 @@ public boolean connect(Predicate<Switch> isTypeSwitchToOperate) {
...
 
-        // Disconnect all switches on node-breaker terminals
-        switchForDisconnection.forEach(sw -> sw.setOpen(false));
-        return isNowConnected;
+        return ConnectDisconnectUtil.connectAllTerminals(
+            this,
+            terminalsToConnect,
+            isTypeSwitchToOperate,
+            this.getNetwork().getReportNodeContext().getReportNode());
```
Comment (flo-dup) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractConnectable.java:214]: You can remove the `this` here (and in the other calls to connect/disconnect)
```suggestion
            getNetwork().getReportNodeContext().getReportNode());
```

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/HvdcLineImpl.java [hunk 6]
```diff
@@ -231,6 +235,51 @@ public void remove() {
...
+        return connectConverterStations(SwitchPredicates.IS_NONFICTIONAL_BREAKER, null);
+    }
+
+    @Override
+    public boolean connectConverterStations(Predicate<Switch> isTypeSwitchToOperate) {
+        return connectConverterStations(isTypeSwitchToOperate, null);
+    }
+
+    @Override
+    public boolean connectConverterStations(Predicate<Switch> isTypeSwitchToOperate, TwoSides side) {
```
Comment (flo-dup) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/HvdcLineImpl.java:249]: Shouldn't we rename this to `connectConverterStation`? It's strange to give a side and have the plural form, and also to interpret the null value as both sides.

Comment (rolnico) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/HvdcLineImpl.java:249]: For the null value, it's already like this for the connectables so I kept the same logic. Same for the plural form : it's `getTerminals(ThreeSides side)` even when we give a ThreeSide so I tried to keep it coherent

**Review by olperr1 — CHANGES_REQUESTED** [review]
@annetill: Should we also revert the javadoc changes of #3079?

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractConnectable.java (hunk 5 above)
Comment (olperr1) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractConnectable.java]: Some are still present in `HvdcLineImpl` and `TieLineImpl`.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/ConnectDisconnectUtil.java [hunk 7]
```diff
@@ -28,6 +28,15 @@ public final class ConnectDisconnectUtil {
     private ConnectDisconnectUtil() {
     }
 
+    /**
+     * Connect the specified terminals of the identifiable. It will connect all the specified terminals or none if at
```
Comment (annetill) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/ConnectDisconnectUtil.java:32]: terminals of the identifiable does not sound good. Terminals are related to connectable, exception for HVDC line and tie line only.

**Review by olperr1 — APPROVED** [review]

**Pull request merged**